# DS-SF-34 | 01 | What is Data Science | Assignment | Starter Code

## Python Review

Programming is a must-have skill for data scientists.  Today, to give you some more practice beyond the course pre-requisites, we are going to implement a few functions in Python.  This assignment covers to some extend the following topics:

- Functions (defining and using your own functions but also calling functions from packages)
- Loops
- Arithmetic operations
- Conditional statements

**Don't worry if you get stuck.  Ask around, review the answer key, and ask around more.  As this course progresses, your programming proficiency will increase.**

> ### Question 1.  Multiples of 3 and 5
>
> If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9.  The sum of these multiples is 23.
>
> Find the sum of all the multiples of 3 or 5 below 1,000.
>
> (Source: [Project Euler | Problem 1](https://projecteuler.net/problem=1))

In [1]:
def sum_multiple_3or5(n):
    totl = 0
    for i in range(1,n):
        if (i % 3 == 0) or (i % 5 == 0):
            totl = totl + i
    return totl

sum_multiple_3or5(1000)

233168

Answer: 233168

> ### Question 2.  Estimating square roots
>
> Given a real number $m$, let's define the series $u$ as follow:
> - $u_0 = 1$
> - $u_{n+1} = \frac{u_n ^ 2 + n}{2u_m}$
>
>
> Implement the calculations of the series $u$ above to estimate square roots.  Verify that $\sqrt{144} = 12$ and use your function to calculate $\sqrt{1024}$.

In [2]:
def sqrt(m):
    u = 1. #u_0 then u_n
    while True:
        v = (u**2 + m) / (2 * u) #u1 then u_n+1
        if abs(u/v-1)< 0.0001:
            break
        u = v
    return round(v)

sqrt(144)

12.0

In [3]:
sqrt(1024)

32.0

Answer: 32

> ### Question 3.  Prime Numbers
>
> A prime (number) is a natural number greater than 1 that has no positive divisors other than 1 and itself.  ([Wikipedia](https://en.wikipedia.org/wiki/Prime_number))
>
> Calculate all primes below 1,000.  What's their sum?

In [4]:
def is_prime(n):
    for p in range(1, n):
        if p % i == 0:
            return False
    return True

totl = 0
p = 1000
for i in range(2, p):
    if is_prime(i):
        totl += 1

totl

998

Answer: 998

> ### Question 4.  Largest prime factor
>
> The prime factors of 13195 are 5, 7, 13 and 29.
>
> What is the largest prime factor of the number 600851475143?
>
> (Source: [Project Euler | Problem 3](https://projecteuler.net/problem=3))

In [5]:
# TODO

Answer: TODO

> ### Question 5.  Mean
>
> Write a function to calculate the mean (average) of a list.
>
> What's the mean of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?

In [6]:
def mean(*params):
    x = sum(params)/len(params)
    return x

mean(10, 8, 13, 9, 11, 14, 6, 4, 12, 7, 5)

9

Answer: 9

> ### Question 6.  Sample standard deviation
>
> Write a function to calculate the standard deviation of a sample.
>
> Given the sample $x_1, x_2, ..., x_N$, its standard deviation is defined as $s = \sqrt{\frac{1}{N - 1} \sum_{i = 1}^{N} (x_i  - \bar{x})^2}$, with $\bar{x}$ as the sample mean.
>
> What's the standard deviation of the following sample: 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?
>
> ([Wikipedia](https://en.wikipedia.org/wiki/Standard_deviation#Sample_standard_deviation))

In [16]:
def mean(*params):
    x = sum(params)/len(params)
    return x

def stdv(*params):
    l = len(params)
    totl = 0.
    for i in params:
        totl = totl + i
    mean = totl/l
    li = []
    for i in params:
        li.append((i-mean)**2)
    mean_diff = float(sum(li)/len(li))
    stdv = sqrt(mean_diff)
    print stdv
    
stdv(10, 8, 13, 9, 11, 14, 6, 4, 12, 7, 5)

3.0


Answer: 3

> ### Question 7.  Median
>
> Write a function to calculate the median ("middle value") of a list.  ([Wikipedia](https://en.wikipedia.org/wiki/Median))
>
> What's the median of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7, and 5?

In [58]:
import math

def median(*params):
    l = len(params)
    s = sorted(params)
    mid = l/2.0
    s2 = [float(i) for i in s]
    if mid % 2 != 0:
        low = s2[int(math.floor(mid))] 
        high = s2[int(math.ceil(mid))]
        print ((high+low)/2)
    else:
        print s2[int(mid)]

    
    
median(10, 8, 13, 9, 11, 14, 6, 4, 12, 7, 5)

9.5


Answer: 9.5

> ### Question 8.  Mode
>
> Write a function to calculate the mode ("most frequent value") of a list.  ([Wikipedia](https://en.wikipedia.org/wiki/Mode_(statistics)))
>
> What's the mode of 10, 8, 13, 9, 11, 14, 6, 4, 12, 7 and 5?  How about the mode of 8, 8, 8, 8, 8, 8, 19, 8, 8 and 8?

In [106]:
def mode(*params):
    s = sorted(params)
    p = dict([[x,params.count(x)] for x in set(params)])
    print p
    m = max(p.values())
    totl = 0
    for i in p.values(): 
        if i == m: 
            totl = + totl + 1
    print 
    if totl == 1:
        for i, j in p:
            if j == m:
                print i
    else: print "no mode"
        
            #in progress
    
mode(10, 8, 13, 9, 11, 14, 6, 4, 12, 7, 5, 5)

{4: 1, 5: 2, 6: 1, 7: 1, 8: 1, 9: 1, 10: 1, 11: 1, 12: 1, 13: 1, 14: 1}



TypeError: 'int' object is not iterable

Answer: TODO